### ISSUE TITLE
[feature-request] Support for ISC-Kea

### ISSUE THREAD

**Issue by ghost** [webmin/webmin#2118]
Internet Systems Consortium has posted notification that ISC-DHCP is no longer supported as of 2022.
 - https://www.isc.org/blogs/isc-dhcp-eol/

Canonical has announced that ISC-Kea is being made available in Noble Numbat 24.04 as a replacement for ISC-DHCP.
 - https://discourse.ubuntu.com/t/noble-numbat-release-notes/39890#kea-40

Webmin should support the incoming ISC-Kea DHCP server in order to make it "feature-compatible" with upcoming Ubuntu releases. This could/should be a standard module, or a third-party module.

**Comment by jcameron** [comment]
Interesting, I didn't know that! Does ISC-Kea use the same config format as the old ISC-DHCPd ?

**Comment by clontarfx** [comment]
ISC have provided a Kea migration tool (KeaMA) that translates ISC-DHCP to ISC-Kea working configurations.
 - https://dhcp.isc.org/

The tool outputs JSON. This can be tested by;
 - Using a simple DHCP configuration: https://gist.githubusercontent.com/paul-ridgway/e419f7e52258db142326b691d0450be3/raw/b7a6cacb098fbc021054580adb02869f5152f33f/simple-dhcpd.conf

Example output;

```
{
  /// This configuration declares some subnets but has no interfaces-config
  /// Reference Kea #245
  "Dhcp4": {
    "valid-lifetime": 86400,
    "max-valid-lifetime": 86400,
    "authoritative": true,
//  "config": [
//    /// log-facility is not supported
//    /// Please use the KEA_LOGGER_DESTINATION environment variable instead
//    {
//      "name": "log-facility",
//      "code": 44,
//      "value": "local7"
//    }
//  ],
    "subnet4": [
      {
        "id": 1,
        "subnet": "192.168.1.0/24",
        "pools": [
          {
            "pool": "192.168.1.100 - 192.168.1.200"
          }
        ],
        "option-data": [
          {
            "space": "dhcp4",
            "name": "domain-name",
            "code": 15,
            "data": "home.ridgway.io"
          },
          {
            "space": "dhcp4",
            "name": "domain-name-servers",
            "code": 6,
            "data": "192.168.1.50, 8.8.8.8, 8.8.4.4"
          },
          {
            "space": "dhcp4",
            "name": "routers",
            "code": 3,
            "data": "192.168.1.1"
          }
        ]
      }
    ]
  }
}
```

**Comment by swelljoe** [comment]
That doesn't really solve anything for a Webmin module, which needs to read and write the native config file format.

**Comment by clontarfx** [comment]
Granted, the examples were given to answer @jcameron regarding file format.

**Comment by jcameron** [comment]
Yeah if the config file format is different, it will take us a fair bit of work to support the new DHCP server.

**Comment by thefunkygibbon** [comment]
whats happening with this then? support for isc dhcp is long gone and kea is the way forward.  but i don't see any support for it within webmin still.  are there any plans as all i see are closed threads about it

**Comment by DevJupyHUB** [comment]
Yes, it has just become timely, THX

**Comment by iliajie** [comment]
> Yeah if the config file format is different, it will take us a fair bit of work to support the new DHCP server.

Jamie, why would it be difficult? Isn’t the Kea DHCP configuration format typically JSON-based?

**Comment by jcameron** [comment]
I mean the semantics of the config format are different I assume? Unless it's exactly the same structure as the old ISC-DHCPd but converted to JSON?

**Comment by iliajie** [comment]
> I mean the semantics of the config format are different I assume? Unless it's exactly the same structure as the old ISC-DHCPd but converted to JSON?

I have no idea off the top of my head...

**Comment by iliajie** [comment]
…perhaps we could adjust the new JSON semantics to match what the old format was...?

**Comment by NuAngel** [comment]
About 18 months in on this request... I'd love to hear if anyone has looked further in to it. I feel like ISC would likely be willing to share the code for "KeaMA" / allow integration of the [offline Migration Assistant package](https://dhcp.isc.org/offline) with the Webmin team to help ease the process of creating a new module for Webmin.

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
